harness/determined/common/api/_session.py

```diff
@@ -1,27 +1,93 @@
-from typing import Any, Dict, Optional
+import abc
+import json as _json
+from typing import Any, Dict, Optional, Tuple, Union
 
 import requests
 import urllib3
 
+import determined as det
+from determined.common import requests as det_requests
 from determined.common import util
-from determined.common.api import authentication, certs, request
+from determined.common.api import authentication, certs, errors, request
 
 
-class Session:
-    def __init__(
-        self,
-        master: Optional[str],
-        user: Optional[str],
-        auth: Optional[authentication.Authentication],
-        cert: Optional[certs.Cert],
-        max_retries: Optional[urllib3.util.retry.Retry] = None,
-    ) -> None:
-        self._master = master or util.get_default_master_address()
-        self._user = user
-        self._auth = auth
-        self._cert = cert
-        self._max_retries = max_retries
+def _do_request(
+    method: str,
+    host: str,
+    path: str,
+    params: Optional[Dict[str, Any]] = None,
+    json: Any = None,
+    data: Optional[str] = None,
+    headers: Optional[Dict[str, str]] = None,
+    cert: Optional[certs.Cert] = None,
+    timeout: Optional[Union[Tuple, float]] = None,
+    stream: bool = False,
+    max_retries: Optional[urllib3.util.retry.Retry] = None,
+) -> requests.Response:
+    # Allow the json json to come pre-encoded, if we need custom encoding.
+    if json is not None and data is not None:
+        raise ValueError("json and data must not be provided together")
+
+    if json:
+        data = det.util.json_encode(json)
+
+    try:
+        r = det_requests.request(
+            method,
+            request.make_url(host, path),
+            params=params,
+            data=data,
+            headers=headers,
+            verify=cert.bundle if cert else None,
+            stream=stream,
+            timeout=timeout,
+            server_hostname=cert.name if cert else None,
+            max_retries=max_retries,
+        )
+    except requests.exceptions.SSLError:
+        raise
+    except requests.exceptions.ConnectionError as e:
+        raise errors.MasterNotFoundException(str(e))
+    except requests.exceptions.RequestException as e:
+        raise errors.BadRequestException(str(e))
+
+    def _get_error_str(r: requests.models.Response) -> str:
```

Comment:
This reads like your choice because it moved files not because you wrote it originally (... at least as part of this PR), but I think `_do_request` is probably the wrong place for this parsing. I think the division of labor where errors take a `requests.models.Response` and do with them whatever is relevant to their particular error type makes more sense than the controller doing that for them.

But I also get that `ForbiddenException` and `UnauthenticatedException` could be generated free from responses. _Currently_, this is the only place a `ForbiddenException` is raised. `UnauthenticatedException`s are raised in two places in `authentication.py`.
1. To re-cast a `ForbiddenException`. I don't like it, but I guess it's what we gotta do. Regardless, it's not taking anything in its constructor.
2. In `set_active`. Looking at it more closely with a better understanding of what `UnauthenticatedException` does, I don't think this is a good choice.

Reply:
2. set_active:  Ha, you are correct.  Looks like `UnauthenticatedException` originally just meant "you are not authenticated" in general, in [the first commit adding user support to the cli](https://github.com/determined-ai/pedl/pull/3422/files#diff-025700446ebbf8daddb56094ab9172d23c9cc33ee261e8cb3da2016028e2a75cR380).

We only started raising `UnauthenticatedException` for 403 errors [months later](https://github.com/determined-ai/pedl/pull/4147/files#diff-2092beea75f37033854275335ec28658ce532fd5bb7fdabec7da37adf3215832R61-R63), then we added `ForbiddenException` and raised it instead [years later](https://github.com/determined-ai/determined/pull/2870/files#diff-27d7a2266bb26e58ae771f161876589e367c437f19d8107f3d28468bda0e935dR127), and starting raising `UnauthenticatedException` for 401's [another year after tahat](https://github.com/determined-ai/determined/pull/5206/files#diff-27d7a2266bb26e58ae771f161876589e367c437f19d8107f3d28468bda0e935dR165-R169).

So the history of definition of these exceptions is much messier than I thought upon first investigation.

I agree that something is wrong here.  How strongly do you feel about including that change as part of this PR?

Reply:
Resolving and adding to the wiki.